File: SimulateData.py
import pandas as pd
import numpy as np
import random
import matplotlib.pyplot as plt

from FirstStageFunctions import *
from ModelTypes import *
from EquilibriumFunctions import *
from ModelFunctions import *
from EstimationFunctions import *
from ParallelFunctions import *
from Derivatives import * 
from NumericalDerivatives import *

random.seed(2220242381)

#### Data Size ####
M = 2 #Markets
J = 5# Lenders
N = 2000 # Consumers per market


#### Model Parameters

# #### Discount Data
# ffr = [0.024,0.018]
# T = len(ffr)

# D_data = np.zeros((T,T))
# for t in range(T):
#     D_data[t,t] = 1.0

# Interest Rate Wedge
psi = 0.2

#### Lender Demand Data 
X = np.zeros((J,J))

for j in range(J):
    X[j,j] = 1.0


# Lender Demand Parameters (J)
β_x = np.array([9.3,9.1, 8.9, 8.7,8.5])

#### Cost Data
cost_type_num = 7

cost_types = [np.zeros(cost_type_num) for i in range(cost_type_num)]
for i in range(len(cost_types)):
    type_dummy = cost_types[i]
    type_dummy[i] = 1.0
    cost_types[i] = type_dummy



# Cost Parameters
γ_WH = np.linspace(0.01,0.04,cost_type_num)
γ_WS = np.linspace(-0.01,0.01,cost_type_num) + γ_WH 
γ_diff = γ_WS - γ_WH

#### Consumer Data
Z_ind = np.random.choice(range(cost_type_num),M*N)

Z = np.zeros((M*N,cost_type_num))
for i in range(Z.shape[0]):
        # Z[i,j] = cons_bounds[j][1] + random.random()*(cons_bounds[j][1] - cons_bounds[j][0])
        Z[i,] =cost_types[Z_ind[i]]

### Extra Parameters
σ = 0.05 # Balance-sheet shock variance

##### Parameters Uncovered Not Estimated 
α_min = -10 # Mean price sensitivity
α_max = -200 # Price sensitivity dispersion

# Log normal distribution
α_mean = 0 # Mean price sensitivity
α_var = 1 # Price sensitivity dispersion
α_scale = -100


theta = Parameters_Sim(β_x,γ_WH,γ_WS,σ,psi)

## Market Data 
# Market Index
# Lender Index
# Market Characteristics
# Lender Characteristics

market_data = np.zeros((J*M,3 + X.shape[1]))

index = -1
m_ind =-1
for m in range(M):
    m_ind +=1
    index_j = -1 
    for j in range(J):
        index +=1
        index_j +=1
        market_data[index,0] = m_ind
        market_data[index,1] = j
        market_data[index,2:(2+X.shape[1])] = X[index_j,:]
            # market_data[index,(2+X.shape[1]):(2+X.shape[1]+W.shape[1])] = W[index_j,:]
            # market_data[index,(2+X.shape[1]+W.shape[1])] = ffr[index_t]


#### MBS Data ####
# Time Index
# Coupon Prices 0.5 - 15, 0.5 increments

# coupon_rates = np.arange(1,9.1,0.25)/100
# prices = [1, 1.25, 1.5, 1.75, 2, 2.25, 2.5, 2.75, 3, 3.25, 3.5, 3.75, 4, 4.25, 4.5, 4.75, 5, 5.25, 5.5, 5.75, 6, 6.25, 6.5, 6.75, 7, 7.25, 7.5, 7.75, 8, 8.25, 8.5, 8.75, 9]
# prices = [str(p) for p in prices]
# mbs_data = pd.read_csv("mbs_data_real_subset.csv") # MBS coupon price data


#### Outside Share Data ####
# Market Index
share_data = np.zeros((M,5+J))
for t in range(M):
    share_data[t,0] = t




##### Data Structure
## Consumer Choice Data
# Market Index
# Consumer credit risk
# Price Sensitivity
# Chosen Lender Index
# Paid interest Rate
# Loan Sold

consumer_data = np.zeros((M*N,3+Z.shape[1]+20))#+2*D_data.shape[1]))#+W.shape[1]+1+1+1+1+1+1+1+1))

index = -1
m_ind = -1
final_ind = -1 
for m in range(M): 
    # index_t = -1
    # for t in range(T):
    m_ind +=1
    # index_t+=1
    for n in range(N):
        index +=1 
        consumer_data[index,0] = m_ind
        # consumer_data[index,1] = index_t
        # consumer_data[index,2] = α_min + random.random()*(α_max-α_min)
        consumer_data[index,2] = α_min + np.exp(random.gauss(α_mean,α_var))*α_scale
        # consumer_data[index,3] = Chosen Lender
        # consumer_data[index,4] = Interest Rate
        # consumer_data[index,5] = Sold or Not
        # consumer_data[index,6] = ffr[index_t]
        consumer_data[index,7:(7+Z.shape[1])] = Z[index,:]
       # last = 7+Z.shape[1]
        # consumer_data[index,(last):(last+D_data.shape[1])] = D_data[index_t,:]
        #last = last+D_data.shape[1]
        # consumer_data[index,(last):(last+D_data.shape[1])] = D_data[index_t,:]
        #final_ind = last+D_data.shape[1]


demand_ind = list(range(2,(2+X.shape[1])))
# demand_ind.append(market_data.shape[1]-1)
# demand_ind.append(0)
# cost_ind = list(range(2+X.shape[1],(2+X.shape[1]+W.shape[1])))


consumer_ind = list(range(7,(7+Z.shape[1])))
# consumer_ind.append(6)
# last = np.max(consumer_ind) +1
# discount_data_ind = range((last),(last+D_data.shape[1]))
# last = np.max(discount_data_ind) +1
# discount_interaction_ind = range((last),(last+D_data.shape[1]))
final_ind = np.max(consumer_ind) + 1

# alpha_buy = np.zeros(consumer_data.shape[0])
# alpha_out = np.zeros(consumer_data.shape[0])

# count = 0
for i in range(consumer_data.shape[0]):
    market = int(consumer_data[i,0])
    market_index = market_data[:,0]==market
    X_i = market_data[market_index,:][:,demand_ind]
    Z_i = consumer_data[i,consumer_ind]
    W_i = np.array([Z_i for i in range(X_i.shape[0])])
    
    dat = Data(i,X_i,W_i,Z_i,0,0,0)

    # mbs_price_i = mbs_data.loc[time,prices].to_numpy()
    # MBS_par = ModelTypes.MBS_Func(coupon_rates,mbs_price_i)
    α = consumer_data[i,2]
    r_eq, itr1,flag = solve_eq_optim(α,dat,theta)
    if not flag:
        print("WARNING: Failed Equilibirum at",i)
        break
    # if α>(-200)
    #     r_eq, itr1 = solve_eq(α,dat,θ,MBS_par)
    # else
    #     r_eq, itr1 = solve_eq_robust(α,dat,θ,MBS_par)
    # end
    # if (α>(-500)) & (itr1>=249)
    #     r_eq, itr1 = solve_eq_robust(α,dat,θ,MBS_par)
    # end



    
    shares = market_shares(r_eq,α,dat,theta)
    sell = balance_sheet_alloc(r_eq,dat,theta)

    share_draw = random.random() #*sum(shares)
    sell_draw = random.random()
    j_select = np.argmax(share_draw<np.cumsum(shares))

    if share_draw > np.sum(shares):
        share_data[market,2] +=1
        consumer_data[i,3] = -1
        # alpha_out[i] = α
        # println("Consumer $i, solve iterations $itr1, inside share: $(sum(shares)) - skip")
        continue
    else:
        # alpha_buy[i] = α
        share_data[market,1] +=1
        share_data[market,5+j_select] +=1

    share_data[market,3] +=1*(1 - sum(shares))
    share_data[market,2] +=1*(sum(shares))

    # check_profit[i] = expected_sale_profit(r_eq,dat,theta,MBS_par)[j_select]
    
    consumer_data[i,3] = j_select
    consumer_data[i,4] = r_eq[j_select]
    consumer_data[i,5] = 1.0*(sell_draw<sell[j_select])
    consumer_data[i,(final_ind + 1):(final_ind + 1 + W_i.shape[1])] = W_i[j_select,:]
    consumer_data[i,final_ind + 1 +  W_i.shape[1]] = 1 - sum(shares)
    consumer_data[i,final_ind + 1 +  W_i.shape[1]+1] = i
    # α_est, r_eq, itr2 = solve_eq_r(r_eq[j_select],j_select,dat,θ,MBS_par)
    # consumer_data[i,7] = α_est

consumer_data = consumer_data[consumer_data[:,3]>=0,:]


np.median(consumer_data[:,4])
np.quantile(consumer_data[:,4],0.1)
np.quantile(consumer_data[:,4],0.9)
np.mean(consumer_data[:,5])

print(len(consumer_data))
print(np.mean(consumer_data[:,5]))

share_data[:,4] = (share_data[:,1] + share_data[:,2])
share_data[:,3] = share_data[:,2]/share_data[:,4]
print(share_data[:,3])

# share_data[0,3] = np.mean(consumer_data[consumer_data[:,1]==0,22])
# share_data[1,3] = np.mean(consumer_data[consumer_data[:,1]==1,22])

file = "consumer_data_sim.csv"
pd.DataFrame(consumer_data).to_csv(file,index=False)

file = "market_data_sim.csv"
pd.DataFrame(market_data).to_csv(file,index=False)

# file = "mbs_data.csv"
# CSV.write(file,DataFrame(mbs_data,:auto))

file = "share_data_sim.csv"
pd.DataFrame(share_data).to_csv(file,index=False)

File: julia/SimulateData.jl
using Random
using LinearAlgebra
using Statistics
using Optim
using Distributions
using Plots
using CSV
using DataFrames
include("halton.jl")
include("types_logit.jl")
include("functions_logit.jl")
include("equilibrium_logit.jl")
include("illustration_logit.jl")


Random.seed!(222024238)

#### Data Size ####
M = 1 #Markets
J = 5# Lenders
N = 1000 # Consumers per market


#### Model Parameters

#### Discount Data
ffr = [0.035,0.02]
T = length(ffr)

D_data = Matrix{Float64}(undef,T,T)
D_data[:].= 0.0
for t in 1:T
    D_data[t,t] = 1.0
end

# Discount Parameters
β_d = [0.5,0.7]

#### Lender Demand Data 
X = Matrix{Float64}(undef,J,J)
X[:].=0.0
for j in 1:J
    X[j,j] = 1.0
end

# Lender Demand Parameters (J)
β_x = [12.9,12.6, 12.4, 12.2,12.0]

#### Cost Data
cost_bounds = [[0,5],[1,100],[0,10]]
W = Matrix{Float64}(undef,J,length(cost_bounds))
for i in 1:size(W,1)
    for j in 1:size(W,2)
        W[i,j] = cost_bounds[j][1] + rand()*(cost_bounds[j][2] - cost_bounds[j][1])
    end
end

# Cost Parameters
ϵ1 = (-0.1 - 0.05)/0.05 # -3.0
# γ_WH = [1.0, 1.0, 1.1, 1.1, 0.9, -3.0]
# γ_WS = [-0.05, -0.0,-0.1,-0.1,-0.2, 3.0]
γ_WH = [0.1, 0.1, 0.2, 0.2, 0.9, -3.0]
γ_WS = [0.15, 0.2,0.1,0.1,-0.3, 3.0]

#### Consumer Data
cons_bounds = [[1,1],[620,850],[0,500],[30,95],[0,20],[0,1],[40,600]]
Z = Matrix{Float64}(undef,M*T*N,length(bounds))
for i in 1:size(Z,1)
    for j in 1:size(Z,2)
        Z[i,j] = cons_bounds[j][1] + rand()*(cons_bounds[j][2] - cons_bounds[j][1])
    end
end

γ_ZH = [0.0,0.0,0.0,0.04,0,-0.05,-0.01]
γ_ZS = [-0.29,0.0,0.0,-0.04,0,0.05,0.01]


### Extra Parameters
σ = 0.05 # Balance-sheet shock variance

##### Parameters Uncovered Not Estimated 
α_min = -60 # Mean price sensitivity
α_max = -600 # Price sensitivity dispersion

m0_mbs = 0.0201 # MBS discount Rate
m1_mbs = (0.06-0.0201)/0.05 # MBS discount slope


θ = Parameters(β_d,β_x,γ_WH,γ_WS,γ_ZH,γ_ZS,σ)


## Market Data 
# Market Index
# Lender Index
# Market Characteristics
# Lender Characteristics

market_data = Matrix{Float64}(undef,J*T*M,2 + size(X,2)+size(W,2))
market_data[:].=0.0

index = 0
m_ind = 0
for m in 1:M
    index_t = 0
    for t in 1:T
        m_ind +=1
        index_t +=1
        index_j = 0 
        for j in 1:J
            index +=1
            index_j +=1
            market_data[index,1] = m_ind
            market_data[index,2] = j
            market_data[index,3:(2+size(X,2))] = X[index_j,:]
            market_data[index,(2+size(X,2)+1):(2+size(X,2)+size(W,2))] = W[index_j,:]
            market_data[index,(2+size(X,2)+size(W,2))] = ffr[index_t]
        end
    end
end


#### MBS Data ####
# Time Index
# Coupon Prices 0.5 - 15, 0.5 increments

coupon_rates = collect(0.5:0.5:15.0)./100
mbs_data = Matrix{Float64}(undef,T,1 + length(coupon_rates))

for t in 1:T
    MBS_par = D_model[t,:]'*[m0_mbs, m1_mbs]
    MBS_par = MBS(MBS_par)
    mbs_data[t,1] = t-1
    mbs_data[t,2:(1+length(coupon_rates))] = P.(coupon_rates,Ref(MBS_par))
end

#### Outside Share Data ####
# Time Index
share_data = Matrix{Float64}(undef,T,5+J)
share_data[:].=0.0
for t in 1:T
    share_data[t,1] = t-1
end



##### Data Structure
## Consumer Choice Data
# Market Index
# Consumer credit risk
# Price Sensitivity
# Chosen Lender Index
# Paid interest Rate
# Loan Sold

consumer_data = Matrix{Float64}(undef,M*T*N,3+size(Z,2)+2*size(D_data,2)+size(D_model,2)+size(W,2)+1+1+1+1+1+1)
consumer_data[:].=0.0



index = 0 
m_ind = 0
final_ind = 0 
for m in 1:M 
    index_t = 0
    for t in 1:T
        m_ind +=1
        index_t+=1
        for n in 1:N
            index +=1 
            consumer_data[index,1] = m_ind
            consumer_data[index,2] = index_t-1
            consumer_data[index,3] = α_min + rand()*(α_max-α_min)
            
            # consumer_data[index,4] = Chosen Lender
            # consumer_data[index,5] = Interest Rate
            # consumer_data[index,6] = Sold or Not
            consumer_data[index,7:(6+size(Z,2))] = Z[index,:]
            last = 6+size(Z,2)
            consumer_data[index,(last+1):(last+size(D_model,2))] = D_model[index_t,:]
            last = last+size(D_model,2)
            consumer_data[index,(last+1):(last+size(D_data,2))] = D_data[index_t,:]
            last = last+size(D_data,2)
            consumer_data[index,(last+1):(last+size(D_data,2))] = D_data[index_t,:]
            final_ind = last+size(D_data,2)
        end
    end
end


demand_ind = Int.(3:(2+size(X,2)))
cost_ind = Int.((2+size(X,2)+1):(2+size(X,2)+size(W,2)))

consumer_ind = Int.(7:(6+size(Z,2)))
last = maximum(consumer_ind)
discount_ind = Int.((last+1):(last+size(D_model,2)))
last = maximum(discount_ind)
discount_data_ind = Int.((last+1):(last+size(D_data,2)))
last = maximum(discount_data_ind)
discount_interaction_ind = Int.((last+1):(last+size(D_data,2)))
final_ind = maximum(discount_interaction_ind)

for i in 1:size(consumer_data,1)
    market = consumer_data[i,1]
    time = Int(consumer_data[i,2])
    market_index = market_data[:,1].==market
    X_i = market_data[market_index,demand_ind]
    W_i = market_data[market_index,cost_ind]
    D_i = consumer_data[i,discount_ind]
    Z_i = consumer_data[i,consumer_ind]
    dat = Data(X_i,W_i,D_i,Z_i)

    MBS_par = D_i'*[m0_mbs, m1_mbs]
    MBS_par = MBS(MBS_par)

    α = consumer_data[i,3]
    r_eq, itr1 = solve_eq_robust(α,dat,θ,MBS_par)

    # if α>(-200)
    #     r_eq, itr1 = solve_eq(α,dat,θ,MBS_par)
    # else
    #     r_eq, itr1 = solve_eq_robust(α,dat,θ,MBS_par)
    # end
    # if (α>(-500)) & (itr1>=249)
    #     r_eq, itr1 = solve_eq_robust(α,dat,θ,MBS_par)
    # end
    shares = market_shares(r_eq,α,dat,θ)
    sell = balance_sheet_alloc(r_eq,α,dat,θ,MBS_par)

    share_draw = rand()#*sum(shares)
    sell_draw = rand()
    j_select = findfirst(share_draw.<cumsum(shares))
    if isnothing(j_select)
        share_data[time+1,3] +=1
        consumer_data[i,4] = -1
        # println("Consumer $i, solve iterations $itr1, inside share: $(sum(shares)) - skip")
        continue
    else
        share_data[time+1,2] +=1
        share_data[time+1,5+j_select] +=1
    end 
    # share_data[time+1,3] +=1*(1 - sum(shares))
    # share_data[time+1,2] +=1*(sum(shares))

    consumer_data[i,4] = j_select-1
    consumer_data[i,5] = r_eq[j_select]
    consumer_data[i,6] = 1.0*(sell_draw.<sell[j_select])
    consumer_data[i,discount_interaction_ind] = consumer_data[i,discount_interaction_ind].*r_eq[j_select]
    consumer_data[i,final_ind + 1] = P(r_eq[j_select] - 0.0025,MBS_par)
    consumer_data[i,(final_ind + 2):(final_ind + 1 + size(W,2))] = W_i[j_select,:]
    consumer_data[i,final_ind + 1 + size(W,2)+1] = 1 - sum(shares)
    consumer_data[i,final_ind + 1 + size(W,2)+2] = i
    # α_est, r_eq, itr2 = solve_eq_r(r_eq[j_select],j_select,dat,θ,MBS_par)
    # consumer_data[i,7] = α_est
    # println("Consumer $i, solve iterations $itr1, inside share: $(sum(shares))")
end

consumer_data = consumer_data[consumer_data[:,4].>=0,:]
println(size(consumer_data)[1])
println(mean(consumer_data[:,6]))

share_data[:,5] = (share_data[:,2] .+ share_data[:,3])
share_data[:,4] = share_data[:,3]./share_data[:,5]


file = "consumer_data.csv"
CSV.write(file,DataFrame(consumer_data,:auto))

file = "market_data.csv"
CSV.write(file,DataFrame(market_data,:auto))

file = "mbs_data.csv"
CSV.write(file,DataFrame(mbs_data,:auto))

file = "share_data.csv"
CSV.write(file,DataFrame(share_data,:auto))

